### PR TITLE
#172 Support for image extra options

### PR DESCRIPTION
**Pull request by NFSS10**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-design/issues/172 |
| Dependencies | https://github.com/ripe-tech/ripe-sdk/pull/213 |
| Decisions | • Added props to ripe-image so it can support the newly added options that were introduced in https://github.com/ripe-tech/ripe-sdk/pull/213 <br>• Added documentation based on https://docs.platforme.com/#render-endpoints-compose and https://github.com/ripe-tech/plugin-base/blob/master/templates/config/applier.html.tpl |
| Screenshot | **Mirrored and rotated ripe-image example:**<br>imagem<br><br>**Example of test toggling mirror and changing rotation every 500ms:**<br>test<br><br>**Updated documentation:**<br>docs3 |

### CODE REVIEW

**Review by gcandal — CHANGES_REQUESTED** [review]
I understand the reasoning but this approach is not an option. If we started `size`/`crop`/`format` as separate props then this must use the same strategy.

**Comment by NFSS10** [comment]
> 
> 
> I understand the reasoning but this approach is not an option. If we started `size`/`crop`/`format` as separate props then this must use the same strategy.

Just to make sure, add all props with comments and examples right?

**Comment by gcandal** [comment]
> > I understand the reasoning but this approach is not an option. If we started `size`/`crop`/`format` as separate props then this must use the same strategy.
> 
> Just to make sure, add all props with comments and examples right?

Comments are enough I guess. What kind of examples were you thinking of? Should be added to the storybook as well.

**Comment by NFSS10** [comment]
> 
> 
> > > I understand the reasoning but this approach is not an option. If we started `size`/`crop`/`format` as separate props then this must use the same strategy.
> > 
> > 
> > Just to make sure, add all props with comments and examples right?
> 
> Comments are enough I guess. What kind of examples were you thinking of? Should be added to the storybook as well.

I was talking about examples in the comments, like we do in other components.
I asked because some options, like `bounding_box` are lists or other types of structures, a little more abstract so I think that atleast those should have an example in the comments on how to use them.

**Comment by gcandal** [comment]
> > > > I understand the reasoning but this approach is not an option. If we started `size`/`crop`/`format` as separate props then this must use the same strategy.
> > > 
> > > 
> > > Just to make sure, add all props with comments and examples right?
> > 
> > 
> > Comments are enough I guess. What kind of examples were you thinking of? Should be added to the storybook as well.
> 
> I was talking about examples in the comments, like we do in other components.
> I asked because some options, like `bounding_box` are lists or other types of structures, a little more abstract so I think that atleast those should have an example in the comments on how to use them.

Agree. It makes sense for those, no doubt.

File: vue/components/organisms/ripe-image/ripe-image.vue [hunk 1]
```diff
@@ -216,7 +567,7 @@ export const RipeImage = {
         },
         imageProps: {
             handler: async function(value) {
-                await this.image.updateOptions(value);
+                await this.image?.updateOptions(value);
```
Comment (gcandal) [inline vue/components/organisms/ripe-image/ripe-image.vue:581]: If this is a problem then all `watch`ers must be adapted the same way, no?

File: vue/components/organisms/ripe-image/ripe-image.vue [hunk 2]
```diff
@@ -257,7 +654,53 @@ export const RipeImage = {
...
+            initialsBuilder: this.initialsBuilder,
+            rotation: this.rotation,
+            flip: this.flip,
+            mirror: this.mirror,
+            boundingBox: this.boundingBox,
+            algorithm: this.algorithm,
+            background: this.background,
+            engine: this.engine,
+            initialsProfile: this.initialsProfile,
+            initialsProfiles: this.initialsProfiles,
```
Comment (gcandal) [inline vue/components/organisms/ripe-image/ripe-image.vue]: Adapt according to https://github.com/ripe-tech/ripe-sdk/pull/213/files#r522188738

File: vue/components/organisms/ripe-image/ripe-image.vue [hunk 3]
```diff
@@ -113,6 +113,357 @@ export const RipeImage = {
...
+        },
+        /**
+         * Line break, is optional and can have one of (normal and word_break).
+         */
+        lineBreaking: {
+            type: Boolean,
+            default: null
+        },
+        /**
+         * Overrides the initialsProfile's shadow, which defines if the initials have a shadow.
```
Comment (gcandal) [inline vue/components/organisms/ripe-image/ripe-image.vue]: *profile's

File: vue/components/organisms/ripe-image/ripe-image.vue [hunk 4]
```diff
@@ -113,6 +113,357 @@ export const RipeImage = {
...
+        },
+        /**
+         * Overrides the initialsProfile's offset to be applied on the shadow.
+         */
+        shadowOffset: {
+            type: String,
+            default: null
+        },
+        /**
+         * Offsets. Example:
```
Comment (gcandal) [inline vue/components/organisms/ripe-image/ripe-image.vue]: Offset of what?

Comment (gcandal) [reply vue/components/organisms/ripe-image/ripe-image.vue]: I guess that if this was copied from somewhere we should update it everywhere with a proper explanation :/

**Review by gcandal — CHANGES_REQUESTED** [review]

File: vue/components/organisms/ripe-image/ripe-image.vue (hunk 4 above)
Comment (NFSS10) [reply vue/components/organisms/ripe-image/ripe-image.vue]: Yeah, kinda thought the same, but as it was like this in https://github.com/ripe-tech/plugin-base/blob/master/templates/config/applier.html.tpl I let it like that

Comment (gcandal) [reply vue/components/organisms/ripe-image/ripe-image.vue]: Go for the 2 for the price of 1 fix 💪

**Comment by gcandal** [comment]
This is waiting on docs and backporting to https://github.com/ripe-tech/ripe-sdk-components-react/

**Review by joamag — APPROVED** [review]

File: vue/components/organisms/ripe-image/ripe-image.vue [hunk 5]
```diff
@@ -706,11 +710,6 @@ export const RipeImage = {
         this.image.update(this.state);
     },
     methods: {
-        setInitials(initials, profiles) {
-            const _initials = initials || "$empty";
-            const _profiles = profiles || null;
-            this.ripeData.setInitials(_initials, _profiles);
-        },
```
Comment (NFSS10) [inline vue/components/organisms/ripe-image/ripe-image.vue]: @gcandal The removal of this method breaks the component behaviour.
Also, I have a question. `profiles` and `engraving` should do the same thing which is update the initials by calling `setInitials` right? But as they are two different contexts they should be both kept as props right? profiles as an `array` and engraving as a `string`

Comment (gcandal) [reply vue/components/organisms/ripe-image/ripe-image.vue]: Conclusion: use some kind of initialsBuilder in WYSYWIG.

```
_defaultInitialsBuilder(initials, engraving, element) {
            return {
                initials: initials || "$empty",
                profile: PROFILES_INPUT_VALUE
            };
        },
```

**Pull request merged**